pinot-broker/src/main/java/org/apache/pinot/broker/routing/timeboundary/TimeBoundaryManager.java

```diff
@@ -142,17 +153,42 @@ private long extractEndTimeMsFromSegmentZKMetadataZNRecord(String segment, @Null
     return endTimeMs;
   }
 
-  private void updateTimeBoundaryInfo(long maxEndTimeMs) {
-    if (maxEndTimeMs > 0) {
-      String timeBoundary = _timeFormatSpec.fromMillisToFormat(maxEndTimeMs - _timeOffsetMs);
-      TimeBoundaryInfo currentTimeBoundaryInfo = _timeBoundaryInfo;
-      if (currentTimeBoundaryInfo == null || !currentTimeBoundaryInfo.getTimeValue().equals(timeBoundary)) {
-        _timeBoundaryInfo = new TimeBoundaryInfo(_timeColumn, timeBoundary);
-        LOGGER.info("Updated time boundary to: {} for table: {}", timeBoundary, _offlineTableName);
+  private synchronized void updateTimeBoundaryInfo(Long enforcedTimeBoundary, long maxEndTimeMs,
+      boolean idealStateReffered) {
+    TimeBoundaryInfo currentTimeBoundaryInfo = _timeBoundaryInfo;
+    Long finalTimeBoundaryMs = null;
+    boolean isEnforced = false;
+    boolean validTimeBoundaryFound = false;
+
+    if (enforcedTimeBoundary != null) {
```

Comment:
This part is quite complicated. It will be good if we can simplify the logic, or put some comments explaining the logic under each if branch

Reply:
I couldn't seem to find a way to simplify this without using 2 different TimeBoundaryInfo instances for explicitly set and derived time boundaries. Have added comments for now.